ci: add basic integration workflow

d3adb5 opened the pull request:
Add a basic continuous integration workflow through GitHub Actions to validate any Terraform code that makes its way to the master branch.

Relates to #7.

d3adb5 committed:
ci: add basic integration workflow

Add a basic continuous integration workflow through GitHub Actions to
validate any Terraform code that makes its way to the master branch.

viniciustrainotti commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,36 @@
+        with:
+          terraform_version: <2.0.0
+
+      - name: Initialize Terraform
+        run: terraform init
```
We will go provides credentials aws to permission in bucket S3 backend 

FYI @damews

viniciustrainotti replied:
Secrets creates in `environment: dev`

Please consider change code for way use to example below.

```yaml
environment: AWS
- name: Set AWS credentials
  uses: aws-actions/configure-aws-credentials@v1
  with:
    aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
    aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
    aws-region: ${{ secrets.AWS_REGION }}
```

My example code by [project](https://github.com/viniciustrainotti/learn-githubactions-react/blob/master/.github/workflows/cd.yaml)

d3adb5 replied:
> We will go provides credentials aws to permission in bucket S3 backend
> 
> FYI @damews

Não ficou muito claro o que foi proposto aqui. Entendo que as credenciais serão adicionadas, mas presumi que ainda não era o caso. Não sendo o caso, não me faz sentido usá-las.

Quanto ao ambiente `dev`, [de acordo com a documentação](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#environment-variables), variáveis de ambiente definidas para um environment só são acessíveis por meio do contexto `vars`, não `secrets`. O contexto `secrets` é usado para secrets no geral.

Ainda nesse tópico, porém, nos convém usar o backend S3 ou até mesmo acessar infraestrutura existente em um build de CI sem código Terraform na `master`? Em algum momento futuro, espero, é claro, termos um `terraform plan` feito nas pull requests. Se isto fosse adicionado antes do código Terraform, levaria a uma falha.

O workflow linkado no seu comentário é de _deployment_ contínuo, não de integração. Deployment está fora do escopo desse PR, ou assim que imaginava.

Um último comentário: se usaremos sempre o mesmo backend Terraform, ele pode ser adicionado como um arquivo `.tf` (e.g. `backend.tf`), assim não precisamos fazer menção explícita a ele nos workflows do GitHub Actions.

damews replied:
+1

A adição do backend.tf seria interessante para comparar a infraestrutura existente. Ainda mais para um _pull request_ para a master.

damews commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,36 @@
+name: CI
+
+on:
+  pull_request: {}
+  push:
+    branches:
+      - master
```
I think it should be great to add the `workflow_dispatch` trigger here. It will be helpful if someone wants to run these checks with their own branch.

d3adb5 replied:
Makes sense to me. Added in 3e6a93f.

damews commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,36 @@
+    branches:
+      - master
+
+jobs:
+  tfsec:
+    name: tfsec
+    runs-on: ubuntu-latest
+    timeout-minutes: 5
+
+    steps:
+      - uses: actions/checkout@v3
+      - uses: aquasecurity/tfsec-action@v1.0.3
+        with:
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+
+  terraform:
+    name: Terraform
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
```
What's the purpose of creating separate jobs to perform operations with the same data (the terraform project)? If the `tfsec` job will block the `terraform`, we could do this process in one single job with the `tfsec` as the first step. If there is no dependency on each other and the only thing that concerns us is the failure (project not safe and/or the validation), we could do these jobs in parallel.

d3adb5 replied:
The jobs are not codependent, so GitHub Actions will run them in parallel by default. This behavior is documented in [this overview.](https://docs.github.com/en/actions/using-jobs/using-jobs-in-a-workflow#overview) As to what the purpose may be: not only do these jobs not have any common dependencies or dependencies on one another, they achieve different purposes. For `tfsec`, the purpose is security. For `terraform`, the purpose is to ensure the Terraform code being submitted as a pull request is valid as far as Terraform is concerned.

Later down the line, the `terraform` job will be expanded to include a `terraform plan` invocation, with the added purpose of comparing existing infrastructure to what has been declared. In and of itself, that cannot be considered a status check, as unintended changes may have been planned, but the Terraform plan is a more than valid requirement for a CI build, as it helps reviewers spot what changes would take place after a merge.

damews commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,36 @@
+name: CI
```
_Security Check and Infrastructure Planning_ ?

d3adb5 replied:
Como se trata de um build de CI (fazer validações antes de deixar qualquer coisa passar pra `master`), acho que não faz muito sentido descrever o que tem dentro. Até porque pode mudar no futuro, conforme formos adicionando mais checks e comportamentos. "CI" é breve e vago, mas era isso que estava buscando, mesmo. O que acha?

d3adb5 committed:
ci: add workflow_dispatch trigger

Add the workflow_dispatch trigger to the continuous integration
workflow, allowing contributors to run the workflow manually on their
own branches.
